File: career/frontend/types/skill.ts
// Skill Types - matches skill-service (port 8007) schemas

export interface SkillBase {
  skill_cd: string;
  skill_nm: string;
  skill_nm_en?: string;
  category?: string;
  difficulty?: number; // 1-5
}

export interface SkillResponse extends SkillBase {
  synonyms?: string[];
}

export interface StudentSkillBase {
  skill_cd: string;
  current_level: number; // 1-5
  target_level: number; // 1-5
}

export interface StudentSkillResponse extends StudentSkillBase {
  student_skill_id: string;
  skill_nm?: string;
  evidence_count: number;
  last_verified_date?: string;
  verification_source?: string;
  trend: 'improving' | 'stable' | 'declining';
  gap: number; // target_level - current_level
}

export interface StudentSkillUpdate {
  current_level?: number;
  target_level?: number;
  verification_source?: string;
}

// Role-Skill Mapping
export interface RoleSkillMapResponse {
  role_cd: string;
  role_nm?: string;
  skill_cd: string;
  skill_nm?: string;
  required_level: number;
  importance: 'critical' | 'important' | 'nice_to_have';
  market_demand_score?: number;
  growth_trend: 'growing' | 'stable' | 'declining';
}

// Skill Graph
export interface SkillNode {
  id: string; // skill_cd
  name: string;
  category?: string;
  difficulty?: number;
  student_level?: number;
  required_level?: number;
  gap?: number;
  importance?: string;
}

export interface SkillEdge {
  source: string;
  target: string;
  relation_type: 'prerequisite' | 'complementary' | 'alternative' | 'builds_on';
  strength: number;
}

export interface SkillGraphResponse {
  nodes: SkillNode[];
  edges: SkillEdge[];
  role_cd?: string;
  role_nm?: string;
}

// Gap Analysis
export interface SkillGapItem {
  skill_cd: string;
  skill_nm: string;
  current_level: number;
  required_level: number;
  gap: number;
  importance: string;
  priority_rank: number;
  recommended_actions?: string[];
}

export interface GapAnalysisRequest {
  student_id: string;
  target_role_cd: string;
  include_recommendations?: boolean;
}

export interface GapAnalysisResponse {
  analysis_id?: string;
  student_id: string;
  target_role_cd: string;
  target_role_nm?: string;
  analysis_date: string;
  overall_gap_score: number; // 0-100, lower is better
  readiness_percentage: number; // 0-100, higher is better
  skill_gaps: SkillGapItem[];
  top_priority_skills: string[];
  strengths: string[];
  summary: string;
}

// Skill Relation
export interface SkillRelationResponse {
  skill_cd_from: string;
  skill_nm_from?: string;
  skill_cd_to: string;
  skill_nm_to?: string;
  relation_type: string;
  strength: number;
}

// Role Requirement
export interface RoleRequirementResponse {
  role_cd: string;
  role_nm: string;
  role_nm_en?: string;
  industry?: string;
  required_skills: RoleSkillMapResponse[];
  total_skills: number;
  critical_skills: number;
  average_required_level: number;
}

File: career/services/skill-service/app/schemas/skill.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date
from uuid import UUID


# ============================================
# Skill Schemas
# ============================================

class SkillBase(BaseModel):
    skill_cd: str
    skill_nm: str
    skill_nm_en: Optional[str] = None
    category: Optional[str] = None
    difficulty: Optional[int] = Field(None, ge=1, le=5)


class SkillResponse(SkillBase):
    synonyms: Optional[List[str]] = None

    class Config:
        from_attributes = True


class StudentSkillBase(BaseModel):
    skill_cd: str
    current_level: int = Field(1, ge=1, le=5)
    target_level: int = Field(3, ge=1, le=5)


class StudentSkillResponse(StudentSkillBase):
    student_skill_id: UUID
    skill_nm: Optional[str] = None
    evidence_count: int = 0
    last_verified_date: Optional[date] = None
    verification_source: Optional[str] = None
    trend: str = "stable"
    gap: int = 0  # target_level - current_level

    class Config:
        from_attributes = True


class StudentSkillUpdate(BaseModel):
    current_level: Optional[int] = Field(None, ge=1, le=5)
    target_level: Optional[int] = Field(None, ge=1, le=5)
    verification_source: Optional[str] = None


# ============================================
# Role-Skill Mapping Schemas
# ============================================

class RoleSkillMapResponse(BaseModel):
    role_cd: str
    role_nm: Optional[str] = None
    skill_cd: str
    skill_nm: Optional[str] = None
    required_level: int
    importance: str
    market_demand_score: Optional[float] = None
    growth_trend: str = "stable"

    class Config:
        from_attributes = True


# ============================================
# Skill Graph Schemas
# ============================================

class SkillNode(BaseModel):
    """Graph node representing a skill"""
    id: str  # skill_cd
    name: str
    category: Optional[str] = None
    difficulty: Optional[int] = None
    student_level: Optional[int] = None  # 학생 현재 레벨
    required_level: Optional[int] = None  # 목표 역할 요구 레벨
    gap: Optional[int] = None
    importance: Optional[str] = None


class SkillEdge(BaseModel):
    """Graph edge representing skill relationship"""
    source: str  # skill_cd_from
    target: str  # skill_cd_to
    relation_type: str  # prerequisite, complementary, alternative, builds_on
    strength: float = 1.0


class SkillGraphResponse(BaseModel):
    """Complete skill graph for visualization"""
    nodes: List[SkillNode]
    edges: List[SkillEdge]
    role_cd: Optional[str] = None
    role_nm: Optional[str] = None


# ============================================
# Gap Analysis Schemas
# ============================================

class SkillGapItem(BaseModel):
    skill_cd: str
    skill_nm: str
    current_level: int
    required_level: int
    gap: int
    importance: str
    priority_rank: int
    recommended_actions: Optional[List[str]] = None


class GapAnalysisRequest(BaseModel):
    student_id: str
    target_role_cd: str
    include_recommendations: bool = True


class GapAnalysisResponse(BaseModel):
    analysis_id: Optional[UUID] = None
    student_id: str
    target_role_cd: str
    target_role_nm: Optional[str] = None
    analysis_date: datetime
    overall_gap_score: float  # 0-100, 낮을수록 좋음
    readiness_percentage: float  # 0-100, 높을수록 좋음
    skill_gaps: List[SkillGapItem]
    top_priority_skills: List[str]
    strengths: List[str]  # 이미 충족된 스킬
    summary: str

    class Config:
        from_attributes = True


# ============================================
# Skill Relation Schemas
# ============================================

class SkillRelationResponse(BaseModel):
    skill_cd_from: str
    skill_nm_from: Optional[str] = None
    skill_cd_to: str
    skill_nm_to: Optional[str] = None
    relation_type: str
    strength: float

    class Config:
        from_attributes = True


# ============================================
# Role Requirement Schemas
# ============================================

class RoleRequirementResponse(BaseModel):
    role_cd: str
    role_nm: str
    role_nm_en: Optional[str] = None
    industry: Optional[str] = None
    required_skills: List[RoleSkillMapResponse]
    total_skills: int
    critical_skills: int
    average_required_level: float

    class Config:
        from_attributes = True
